src/net/send_recv/msg.rs

```diff
@@ -591,6 +597,75 @@ impl<'buf> Iterator for AncillaryDrain<'buf> {
 
 impl FusedIterator for AncillaryDrain<'_> {}
 
+/// An ABI-compatible wrapper for `mmsghdr`, for sending multiple messages with
+/// [sendmmsg].
+#[cfg(target_os = "linux")]
+#[repr(transparent)]
+pub struct MMsgHdr<'a> {
+    raw: c::mmsghdr,
+    _phantom: PhantomData<&'a mut ()>,
+}
+
+#[cfg(target_os = "linux")]
+impl<'a> MMsgHdr<'a> {
+    /// Constructs a new message with no destination address.
+    pub fn new(iov: &[IoSlice<'a>], control: &mut SendAncillaryBuffer<'_, '_, '_>) -> Self {
+        with_noaddr_msghdr(iov, control, Self::wrap)
+    }
+
+    /// Constructs a new message to a specific IPv4 address.
+    pub fn new_v4(
+        addr: &SocketAddrV4,
+        iov: &[IoSlice<'a>],
+        control: &mut SendAncillaryBuffer<'_, '_, '_>,
+    ) -> Self {
+        with_v4_msghdr(addr, iov, control, Self::wrap)
```

Comment:
Isn't this unsound? `with_v4_msghdr` allocates the V4 socket address on the stack and then passes that stack address in the `msghdr` in `Self::wrap`. So when we eventually call into the system call it is pointing to invalid stack memory.

Reply:
Yes, you're right. Maybe `SockAddrStorage` can be used for this?

Reply:
In order to avoid passing the length in here, we'll need to add a `RawSockAddr` type, I think, created with `SockAddrV4::as_raw` etc. It'll be slightly unergonomic, as the borrow checker won't let you do this:

```rust
MMsgHdr::new_with_addr(my_v4_addr.as_raw(), ...)
```

Since that would create a temporary value which is dropped. But it should work fine otherwise.

Reply:
Add a `RawSocketAddr` type in my latest push (it's a separate commit, so can be reviewed independently).